patroni/postgresql.py

```diff
@@ -707,7 +707,20 @@ def is_running(self):
         if not (self._version_file_exists() and os.path.isfile(self._postmaster_pid)):
             # XXX: This is dangerous in case somebody deletes the data directory while PostgreSQL is still running.
             return False
-        return self.is_pid_running(self.get_pid())
+        pidfile = self.read_pid_file()
```

Comment:
is there any reason for not moving the new checks inside `is_pid_running()` instead of adding them to `is_running()`?

Reply:
I can confirm this is also required in `is_pid_running` to avoid getting stuck in `patroni/postgresql.py#Postgresql._wait_for_postmaster_stop` when some other existing process happens to have the same PID as written in a now defunct postmaster's `postmaster.pid` file.

Reply:
Not a really good reason. By doing it in is_running() I avoided reading the pid file twice. As far as I can tell the places where is_pid_running() is called currently don't care about this specific failure case. When starting we already know we have the correct pid, when stopping we first check for the correct process via is_running() and when aborting a restart postmaster was supposed to be running before restart was initiated so the pid file should have been correct.

I could rejiggle the code by adding an optional start_time parameter to is_pid_running() if you think that is nicer.

We discussed this with @CyberDem0n at pgconf.eu. I had concerns about the start time check possibly not being robust in face of system clock changes. He proposed that we look at the command name instead. That would fail when somebody runs a postgres instance as the same user (otherwise kill -0 permission check fails), but in a different data directory. I'll test how start time behaves on Linux if system clock is screwed with and will report back.

Reply:
@ants indeed, something feels strange.

#548 tells that async executor stuck in `_wait_for_postmaster_stop`, but this PR does not really change it and `_wait_for_postmaster_stop` still calls `is_pid_running`.

I am pretty sure, it will fix exactly your problem, when you start patroni after power loss, but it will not fix another problem when OS will create a new process with the same PID as postmaster had.

And regarding process name check: we always start postmaster by calling "postgres" binary.
In this case process name always matches with name of the binary (at least on linux).

P.S. can you also have a look on https://github.com/zalando/patroni/pull/553 ?

Reply:
`_wait_for_postmaster_stop` is called from `_do_stop`, which does `is_running()` check at the beginning, catching the stale pid file case. You're right that there is still a race condition if postgres terminates while we are waiting and kernel reassigns its pid to a process of the same user within the polling interval. I think moving the start time check into `is_pid_running()` is probably a good idea. I will do that.

With the name check I'm now not worried about false negative (postgres running as something else than postgres), but about false positive (postgres that is running is not "our" postgres).

Reply:
I think it is a good idea to check for the process name in addition to the timestamp, and also do this inside `_is_pid_running`. I'd go even further and rename the function to `_is_postmaster_pid_running`, since it is only used for the checking the postmaster pid so far.  We can indeed pass the process start time from the code that has read the pid originally to that function (so far it seems the only place where we don't immediately pass the pid to the `_is_pid_running` is s `wait_for_port_open`), and assume the postmaster process name remains unchanged, so that we can make it a constant.